arthurcro/default diagnostic readability improvements

arthurcro opened the pull request:
<!--
If you're opening a PR to cherry-pick a change for a release branch, use this template instead:
https://github.com/apple/swift-docc/blob/main/.github/PULL_REQUEST_TEMPLATE/CHERRY_PICK.md
-->

Bug/issue #, if applicable: 

## Summary

_Provide a description of what your PR addresses, explaining the expected user experience. 
Also, provide an overview of your implementation._

## Dependencies

_Describe any dependencies this PR might have, such as an associated branch in another repository._

## Testing

_Describe how a reviewer can test the functionality of your PR. Provide test content to test with if
applicable._

Steps:
1. _Provide setup instructions._
2. _Explain in detail how the functionality can be tested._

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ ] Added tests
- [ ] Ran the `./bin/test` script and it succeeded
- [ ] Updated documentation if necessary

arthurcro committed:
Improve default diagnostic formatter

d-ronnqvist commented on Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift:
```diff
@@ -18,38 +18,56 @@ public final class DiagnosticConsoleWriter: DiagnosticFormattingConsumer {
     public func finalize() throws {
-        // The console writer writes each diagnostic as they are received.
+        if formattingOptions.contains(.formatConsoleOutputForTools) {
+            // FIXME:
```
I don't think there's anything else to do here. This comment just provides an explanation to why this branch of the if statement is empty.

arthurcro committed:
Fix suggestion display ordering

arthurcro committed:
Highlight output only when connected to termnial

arthurcro committed:
Highlight diagnostic summary only when connected to terminal

arthurcro committed:
Fix exisiting tests to use tools output formatting option

arthurcro committed:
Add default diagnostic formatting tests

d-ronnqvist committed:
Add successfully resolved external references to reference index (#582)

* Add successfully resolved external references to reference index

rdar://108974747

* Undo accidental whitespace change

Co-authored-by: Ethan Kusters <ekusters@apple.com>

arthurcro committed:
Relocate default diagnostic formatting test to a separate file

arthurcro committed:
Remove URL force unwrapping

arthurcro committed:
Refactor formatted description for diagnostic

arthurcro committed:
Refactor suggestion grouping

arthurcro committed:
Remove redundant ANSIAnnotation color reset

arthurcro committed:
Improve suggestion vertical alignment prefix

arthurcro committed:
Revert public initializer changes

To make it easier to fix the formatter abstraction later, we keep the previous public initializer unchanged and add a new internal initializer that accept all 4 arguments.

arthurcro committed:
Clean up source lines after writing the diagnostics

Since the sourceLines could potentially be big if there were diagnostics in many large files, we remove the cached lines after the console writer finished writing the diagnostics.

arthurcro committed:
Improve inline documentation

arthurcro committed:
Clean up source lines cache after all diagnostics have been formatted

Introduces an API to inform the formatter that all diagnostics have been formatted.

finagolfin committed:
Support building this repo for more platforms, by checking the build triple (#500)

d-ronnqvist committed:
Prefer non-symbols in general documentation links (#594)

* Prefer non-symbols in general documentation links

rdar://109583745

* Elaborate comment about filtering symbol and non-symbol collisions.

Also, add test where  symbol and non-symbol have same path.

compnerd committed:
docc: indicate Windows as being supported

Allow list Windows execution.  Without this change, we would fail to run
the CLI tool.

compnerd committed:
LMDB: introduce a typealias for `mode_t` (#600)

`mode_t` is a non-portable extension to C (POSIX).  Introduce a new
typealias `ModeType` that represents the underlying type that is used.
Provide an alternate definition for the default mode on Windows as it
does not have non-user permissions on a file but rather relies on proper
ACLing.

compnerd committed:
Benchmark: implement memory query for Windows (#598)

Implement support for querying the high water mark for the working set
for the current process.  This is accomplished with the PSAPI (Version
2) API from Windows.

compnerd committed:
build: fix dependencies (#599)

* build: fix dependencies

We need this to use the new initialiser `init(url:exact:)` for updating
the standalone build to match the CI build.

When building in CI, we would build against swift-crypto 2.5.0 as per
update-checkout-config.json in apple/swift.  Adjust the standalone
builds to match.

When building on CI, we would build against swift-argument-parser 1.2.2
as per update-checkout-config.json in apple/swift.  Make the standalone
builds match.

* - Update Package.swift for Swift 5.6 minimum (as already required by NIO).
- Remove deprecated usages and bump the dependency version requirements for NIO, ArgumentParser, and SwiftCrypto (most especially the latter so that updates in the 2.x series are available).
- Remove old #if swift conditionals no longer required
- Package.resolved now uses the version 2 format

* The platform-specific handling for conforming Array to LMDBData is still required.

---------

Co-authored-by: Gwynne Raskind <gwynne@vapor.codes>

compnerd committed:
FoundationExtensions: enable autoreleasepool on Windows (#602)

The `autoreleasepool` is technically required on all non-ObjC runtimes.
It is likely better to use the unstable `_runtime(_ObjC)` check but for
now simply extend the OS list with Windows.

compnerd committed:
Model: enable concurrent rendering on Windows (#601)

Assume that concurrent rendering on Windows is the proper thing to do.
It is unclear why Linux (but not Android) uses serial rendering.  This
enables us to make progress with the Windows port.

compnerd committed:
Utility: add a Windows path to the synchronisation support (#603)

Use a `SRWLOCK` on Windows to support locking.  Slim Reader-Writer locks
are lightweight mutexes that closely resemble `pthread_mutex_t` in
functionality.  While possible to use `WaitOnAddress` to implement
`os_unfair_lock`, that would be more complex and a slightly heavier
mutex as a starting point seems reasonable.  If it turns out that the
performance really requires an even lighter weight locking primitive, we
can refine this to `WaitOnAddress` subsequently.

compnerd committed:
build: remove the PreviewServer on Windows (#604)

This is the critical change to enable Windows support in DocC.  The
dependency on NIO is restricted to the PreviewServer and as NIO is
incapable of supporting non-POSIX models (or emulation of POSIX models),
we cannot use this dependency on Windows.  As a temporary solution,
remove the NIO dependency on Windows by removing the PreviewServer.

d-ronnqvist committed:
Update distribution documentation static environment enhancements (#587)

* Update distribution docs with static hosting and hosting base path info

* Update copyright year in documentation article

d-ronnqvist committed:
Various small documentation phrasing updates (#590)

d-ronnqvist committed:
Update content formatting documentation with symbol link enhancements (#588)

* Update content formatting documentation with symbol link enhancements

* Minor phrasing adjustments about link disambiguation

compnerd committed:
Utility: disable `DirectoryMonitor` on Windows temporarily (#606)

This temporarily disables the functionality on Windows to match the
Linux port.  We should be able to reimplement this functionality as
already done in clang and swift-tools-support-core on Windows.

compnerd committed:
Utility: disable `Signal` on Windows (#605)

Windows does not support the concept of signals and has no equivalent to
it.  Simply disable the interface to ensure that we do not accidentally
rely on it somewhere.

compnerd committed:
build: bump swift-docc-plugin to 1.2.0 for Windows (#607)

The Windows support was added in 1.1.0.  Update to the latest release of
the plugin to support Windows.

compnerd committed:
build: split the SwiftDocCUtilitiesTest for Windows (#608)

We need to replicate the target definition to exclude some files when
building for Windows as they depend on NIO to create the PreviewServer.
Rather than adding the files and conditionally removing the content,
remove thee files from being built.

Kyle-Ye committed:
Fix preview issue on macOS and Linux

compnerd committed:
Update Sources/docc/main.swift

Co-authored-by: Kyle <kyle201817146@gmail.com>

Kyle-Ye committed:
Fix Package.resolved syncing issue with Package.swift (#612)

Kyle-Ye committed:
Remove `#if os` check on Package.swift file (#614)

compnerd committed:
Tests: adjust synchronization tests for Windows (#615)

* Tests: adjust synchronization tests for Windows

Windows does not support pthreads.  Adjust the tests to use the Windows
APIs for locking where the primitive being used is the `SRWLOCK`.

* SwiftDocCTestes: extract a helper to check the lock state

Address feedback from @Kyle-Ye in #615.  This avoids replicating the
logic for checking the lock state which is already subtle due to the
checking of the return value.

d-ronnqvist committed:
Fix a crash when linking to a symbol that doesn't have a page (#617)

rdar://110177531

compnerd committed:
Tests: use `XCTUnwrap` over forced unwrappng (#619)

Failure to unwrap would trigger an assertion failure whereas the use of
`XCTUnwrap` will fail the test case and permit the remainder of the test
suite to be processed.  This allows us to better understand the state of
the test suite coverage when porting.

compnerd committed:
Tests: treat Windows as Linux (#620)

Treat Windows similarly to Linux as the functionality is near similar in
completeness.  While it is possible to support the DirectoryMonitor on
Windows, it is not yet implemented in DocC, so simply treat it as we do
Linux.

compnerd committed:
SwiftDocCUtilitiesTests: add wrappers for `setenv` and `unsetenv` (#610)

The strings need to be converted to UTF16 before being set into the
environment control block.  Introduce helper wrappers for the functions
to allow the conversion to be centralised.

compnerd committed:
Tests: replace `sleep` with `Thread.sleep(forTimeInterval:)` (#616)

`sleep` is not portable as it is not a standard C function.  Use the
Foundation function to provide a single codepath across all the targets.

QuietMisdreavus committed:
keep the original URL when decoding Download/ExternalLocationReference (#621)

rdar://110536969

emilyychenn committed:
Add authoring support for `@TitleHeading` directive (#611)

A new `@TitleHeading` metadata directive for customizing the text of a page-title heading (also known as an eyebrow or kick).

`@TitleHeading` accepts an unnamed parameter containing containing the page-title’s heading text.

    @Metadata {
      @TitleHeading("Release Notes")
    }

This feature has been pitched on the forums here:
https://forums.swift.org/t/supporting-more-types-of-documentation-with-swift-docc/59725#titleheading-10

rdar://110662981

MaxDesiatov committed:
Converter: allow converting a directory w/o `.docc` extension (#585)

When converting a directory that only has Markdown files in it, it feels redundant to rename the directory to have a `.docc` extension just for the sake of rendering those files. It would be great if DocC supported converting any directory regardless of its extension out of the box.

This new behavior is conditional on the newly introduced `--allow-arbitrary-catalog-directories` flag.

ethan-kusters committed:
Make `ExternalLocationReference` url property public and mutable (#627)

Allows clients to access and modify the destination URL for an
`ExternalLocationReference`.

rdar://110803672

compnerd committed:
Tests: add a missed case of Windows exclusion (#629)

This mirrors the Linux and Android paths and excludes this test on
Windows.  With this change (and 1 workaround), we can fully execute the
test suite (though not yet pass it completely).

QuietMisdreavus committed:
`@PageKind` sets the kind for the documentation node (#626)

rdar://110182993

compnerd committed:
Tests: adjust LMDB test harness for Windows (#631)

Ensure that we release the open handles before attempting to clean up
the temporary directory.  This repairs the execution of 11 tests on
Windows, bringing us to a 95% pass rate (excluding `LogHandleTests`).

QuietMisdreavus committed:
don't crash on an extension with an empty declaration (#635)

rdar://110683602

compnerd committed:
Tests: disable the LogHandle tests on Windows (#634)

`FileHandle` transacts in `HANDLE`s on Windows rather than file
descriptors.  This makes it impossible to replace the end point of the
pipe after it has been created as the conversion from a HANDLE to a file
descriptor transfers ownership and as such the FileHandle does not work
with file descriptors at all.  Replicating that behaviour here would
detach the `FileHandle` associated with the `Pipe` and thus cannot be
done inline.

d-ronnqvist committed:
Finalize the diagnostic engine if an error is thrown (#637)

This ensures that any gathered diagnostics are printed before exiting.

rdar://110782381

QuietMisdreavus committed:
deprecate ExternalLocationReference in favor of DownloadReference (#638)

rdar://111121386

d-ronnqvist committed:
False positive warning when curating symbol with special character (#642)

* Fix false-positive warning for curated symbols with special characters

rdar://111227479

* Remove redundant symbol index modification

compnerd committed:
Navigator: close the `LMDB.Environment` on `finalize(_:_:_:)` (#646)

Ensure that we close the LMDB environment on the exit from
`finalize(_:_:_:)` method.  This is important for platforms which do not
permit alteration of directories with open references.  Having the
environment be closed after finalization enables the generated index to
be moved into the final destination on Windows.

Kyle-Ye committed:
Make visitLineBreak return new line instead of space (#633)

* Make visitLineBreak return new line instead of space

* Update test case for more human-readable

emilyychenn committed:
Allow certain directives (#641)

Allow certain directives in documentation comments

rdar://111414385

natikgadzhi committed:
Allow anchor <doc:#heading> within same article (#652)

compnerd committed:
Indexing: add a workaround for clearing the index (#647)

The unit tests will occasionally desire to clean up the generated index
while retaining a reference to the NavigatorIndex instance.  This is
prevented on Windows as the LMDB environment will preserve an open
handle which results in the index not being removed.

Co-authored-by: Kyle <kyle201817146@gmail.com>

d-ronnqvist committed:
Remove the cache based link resolver (#624)

* Remove the cache based link resolver

* Change link resolver property to implicitly unwrapped optional

* Remove should-use checks in new tests

arthurcro committed:
Improve C import

The previous import solution was working only for Darwin platforms, it needed to be update to support more platforms.

arthurcro committed:
Merge branch 'main' into arthurcro/default-diagnostic-readability-improvements

arthurcro committed:
Revert whitespace change

A trailing whitespace is added in `TestTutorial.tutorial:42`, this is expected and should not have been removed.

arthurcro committed:
Revert "Revert public initializer changes"

This reverts commit 6e1d0379c984526ebcc6ec73a78a97a98bab6b51.
